### DISCUSSION TITLE
How to make a Podcast plugin that supports enclosures?

### DISCUSSION THREAD

**Discussion by schulle4u** [datenstrom/community#188] [Ask a question] [answered]
Currently I'm trying to extend the feed plugin so it can handle basic podcast enclosures. The user should be able to create a shortcut within a (blog) page like this: 

`[podcast url="some_media.mp3" length="00:30:00" type="audio/mpeg"]`

So I created an usual shortcut code:

```
    // Handle page content parsing of custom block
    function onParseContentBlock($page, $name, $text, $shortcut)
    {
        $output = NULL;
        if($name=="podcast" && $shortcut)
        {
            list($url, $length, $type) = $this->yellow->toolbox->getTextArgs($text);
            if(empty($length)) $length = "0";
            if(empty($type)) $type = "audio/mpeg";
            $output = "<enclosure url=".$url." length=".$length." type=".$type." />";
        }
        return $output;
    }
```

What I need to know is how to print the podcast enclosure within the feed.xml and not within page output. Calling the shortcode from the page parsing function doesn't work. Could anybody give me some tipps or am I just doing it completely wrong during this hot weather? :)

**Comment by markseuffert** [comment]
How about this code, it returns an empty string on normal pages.

```
    $output = NULL;
    if($name=="podcast" && $shortcut)
    {
        list($url, $length, $type) = $this->yellow->toolbox->getTextArgs($text);
        if(empty($length)) $length = "0";
        if(empty($type)) $type = "audio/mpeg";
        $pagination = $this->yellow->config->get("contentPagination");
        $feedFile = $this->yellow->config->get("feedFileXml");
        if($this->yellow->page->get("template")=="feed" && $_REQUEST[$pagination]==$feedFile)
        {
            $output = "<enclosure url=\"".htmlspecialchars($url)."\" length=\"".htmlspecialchars($length)."\" type=\"".htmlspecialchars($type)."\" />";
        } else {
            $output = "";
        }
    }
    return $output;
```

Is this what you are looking for? Sunny weather here in Sweden too. 😄

**Comment by schulle4u** [comment]
Thanks for your help, this solves the question how to select the right template. However, I need the output to appear in the feed code: 

```
                    $output .= "<guid isPermaLink=\"false\">".$page->getUrl()."?".$timestamp."</guid>\r\n";
                    $output .= "<description><![CDATA[".$description."]]></description>\r\n";
                    $output .= "Enclosure goes here!\r\n";
                    $output .= "</item>\r\n";
```

How can I call your above code example from there?

Steffen

**Comment by markseuffert** [comment]
Would be best to know your circumstances. Want to explain more or make a test case?

**Comment by schulle4u** [comment]
Actually I'd like to use the blog plugin or just a content folder with markdown pages to create an audio-blog. These pages are like normal blog pages, except for an additional link to a media file (mp3, ogg or whatever). These files should be downloadable from the web page, this is no problem so far. 

The feed plugin however should output a xml file which can be subscribed with [Podcatchers](https://en.wikipedia.org/wiki/List_of_podcatchers). For this case I need to include the link to the media file as enclosure tag in the xml feed. So my thought was to simply copy the feed plugin to a new one named podcast and add this additional functionality. This also would allow me to have more podcasting specific tags such as iTunes metadata and so on, which otherwise could conflict with the normal feed output. 

Hope I could clarify my plans a bit. 

Steffen

**Comment by markseuffert** [comment, marked as answer]
Ah now I get it, sorry my code above was wrong.

According to the [RSS specification](http://www.rssboard.org/rss-profile#element-channel-item-enclosure) `<enclosure>` is an optional sub-element of  `<item>`. This means that you have to modify the feed plugin. I am not sure what would be the best way, but here are some tips. You could check if the current page has meta data `Enclosure` and then create the necessary enclosures. You could add a manual setting at the top of a page like this:

`Enclosure: some_media.mp3, second_one.mp3`

Or you write a plugin that detects all MP3 links and adds meta data automatically to the page. Whatever you do, in the end the (new) feed plugin has to check if a page has enclosures and add them to the XML.

**Comment by schulle4u** [comment]
Thanks Mark. I'll have a deeper look into the yellow API and see how this could be applied to the feed plugin. In theory this should even be easier to implement then my first idea. Will keep you updated and perhaps make a code contribution if something useful comes out of my head. 

Steffen
